src/classkit.cs

```diff
@@ -0,0 +1,329 @@
+//
+// Copyright 2018 Xamarin Inc. All rights reserved.
+//
+
+#if XAMCORE_2_0
+using System;
+using XamCore.Foundation;
+using XamCore.ObjCRuntime;
+using System.Reflection;
+
+namespace XamCore.ClassKit {
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[Native]
+	enum CLSBinaryValueType : nint {
+		TrueFalse = 0,
+		PassFail,
+		YesNo,
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[Native]
+	enum CLSContextType : nint {
+		None = 0,
+		App,
+		Chapter,
+		Section,
+		Level,
+		Page,
+		Task,
+		Challenge,
+		Quiz,
+		Exercise,
+		Lesson,
+		Book,
+		Game,
+		Document,
+		Audio,
+		Video,
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[Native]
+	[ErrorDomain ("CLSErrorCodeDomain")]
+	public enum CLSErrorCode : nint {
+		None = 0,
+		ClassKitUnavailable,
+		InvalidArgument,
+		InvalidModification,
+		AuthorizationDenied,
+		DatabaseInaccessible,
+		Limits,
+		InvalidCreate,
+		InvalidUpdate,
+		PartialFailure,
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	enum CLSContextTopic {
+		[Field ("CLSContextTopicMath")]
+		Math,
+		[Field ("CLSContextTopicScience")]
+		Science,
+		[Field ("CLSContextTopicLiteracyAndWriting")]
+		LiteracyAndWriting,
+		[Field ("CLSContextTopicWorldLanguage")]
+		WorldLanguage,
+		[Field ("CLSContextTopicSocialScience")]
+		SocialScience,
+		[Field ("CLSContextTopicComputerScienceAndEngineering")]
+		ComputerScienceAndEngineering,
+		[Field ("CLSContextTopicArtsAndMusic")]
+		ArtsAndMusic,
+		[Field ("CLSContextTopicHealthAndFitness")]
+		HealthAndFitness,
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[Static]
+	interface CLSErrorUserInfoKeys {
+
+		[Field ("CLSErrorObjectKey")]
+		NSString ObjectKey { get; }
+
+		[Field ("CLSErrorUnderlyingErrorsKey")]
+		NSString UnderlyingErrorsKey { get; }
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	enum CLSPredicateKeyPaths {
+		[Field ("CLSPredicateKeyPathDateCreated")]
+		DateCreated,
+		[Field ("CLSPredicateKeyPathIdentifier")]
+		Identifier,
+		[Field ("CLSPredicateKeyPathTitle")]
+		Title,
+		[Field ("CLSPredicateKeyPathUniversalLinkURL")]
+		UniversalLinkUrl,
+		[Field ("CLSPredicateKeyPathTopic")]
+		Topic,
+		[Field ("CLSPredicateKeyPathParent")]
+		Parent,
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[BaseType (typeof (NSObject))]
+	[DisableDefaultCtor]
+	interface CLSObject : NSSecureCoding {
+
+		[Export ("dateCreated")]
+		NSDate DateCreated { get; }
+
+		[Export ("dateLastModified")]
+		NSDate DateLastModified { get; }
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (11,3)]
+	[BaseType (typeof (CLSObject))]
+	[DisableDefaultCtor]
+	interface CLSActivity {
+
+		[Export ("progress")]
+		double Progress { get; set; }
+
+		[Export ("duration")]
+		double Duration { get; }
+
+		[Export ("primaryActivityItem", ArgumentSemantic.Assign)]
+		CLSActivityItem PrimaryActivityItem { get; set; }
+
+		[Export ("addProgressRangeFromStart:toEnd:")]
+		void AddProgressRange (double start, double end);
+
+		[Export ("addAdditionalActivityItem:")]
+		void AddAdditional (CLSActivityItem activityItem);
```

Comment:
Maybe just `Add`.

I think `AddProgressRange` above is fine because the 2 double parameters are not enough to understand what we're adding but here it's obvious you add an activityItem.

Also "Add" and "Additional" are too similar, it's weird to have them together.

Reply:
I originally had `AddAdditionalActivityItem`, I am inclined to go back to it. My main reason to not have just `Add` is because the container type isn't named as a collection, maybe `AddActivity`?

Reply:
I see that's a good reason. Then I vote for `AddActivity`. Could be `AddAdditionalActivity` but really... we're adding an activity so it's obviously in "addition" (additional) to the others...

Note: in Swift they went for `addAdditionalActivityItem`...
https://developer.apple.com/documentation/classkit/clsactivity/2953065-addadditionalactivityitem

Reply:
> Then I vote for `AddActivity`. Could be `AddAdditionalActivity` but really... we're adding an activity so it's obviously in "addition" (additional) to the others...

We have `CLSActivity` and `CLSActivityItem` objects, so I think `AddActivityItem` makes more sense and it's clearer, agree that `AddAdditional` sounds "recursive" so +1 on omitting the `Additional` part

Reply:
I disagree :smile:

Reading through Apple's documentation (https://developer.apple.com/documentation/classkit/recording_additional_metrics_about_a_completed_task), there seems to be a "primary" activity, and then you can add additional activities as well, which makes the `Additional` part slightly less redundant (it also clearly states that it will never be a primary activity).

So I'd go for `AddAdditionalActivity` (there's also an `AdditionalActivityItems` property; and whatever we do those two should match).

Reply:
Good point @rolfbjarne, I think I will go for the [swift name](https://developer.apple.com/documentation/classkit/clsactivity/2953065-addadditionalactivityitem) here `AddAdditionalActivityItem` but yeah makes sense now reading it that way, thanks for having a look.